### PR TITLE
Exclude PR refs from mirror push

### LINKED ISSUE #25: Option to ignore refs push error
There are some refs like refs to pull requests that cannot be pushed and will cause the task to fail. Is there any chance we can ignore it and set the task result to success?

Example log:

```
2018-11-05T02:48:25.3673741Z [command]/usr/bin/git -c http.sslVerify=true clone --mirror https://example.com/example.git
2018-11-05T02:48:25.3753874Z Cloning into bare repository 'example.git'...
2018-11-05T02:48:29.3713629Z [command]/usr/bin/git -c http.sslVerify=true -C example.git push --mirror https://example.com/example.mirror.git
2018-11-05T02:48:31.2188400Z To https://example.com/example.mirror.git
2018-11-05T02:48:31.2189754Z  * [new branch]      master -> master
2018-11-05T02:48:31.2191003Z  ! [remote rejected] refs/pull/5/head -> refs/pull/5/head (The current action can only be performed by the system.)
2018-11-05T02:48:31.2191437Z  ! [remote rejected] refs/pull/6/head -> refs/pull/6/head (The current action can only be performed by the system.)
2018-11-05T02:48:31.2191864Z error: failed to push some refs to 'https://example.com/example.mirror.git'
2018-11-05T02:48:31.2251524Z ##[error]Error: /usr/bin/git failed with return code: 1
2018-11-05T02:48:31.2275053Z ##[section]Finishing: Mirror repo
```

### PR DESCRIPTION
**Pull request by calebcartwright**
## Changes
Summary of changes included in this PR:
- [X] Refactors/simplifies unit tests
- [x] Enhance VS Code content 
- [x] Enhances functionality to exclude PR refs from mirror operations

New functionality has been tested, and can be reviewed at [this build pipeline in our public OSS Azure DevOps Project](https://dev.azure.com/swellaby/OpenSource/_build/index?definitionId=41&_a=completed)

## Issues Completed
- Closes #25 

CC - @traviskosarek @calebcartwright

### CODE REVIEW

**Comment by calebcartwright** [comment]
@traviskosarek - will leave this open for a couple days for you to review, and have included some comments below. Since that the functionality is already working/validated though I will go ahead and merge if you don't get a chance so that we can get the fix out. 

Things to note:
* I did some pretty heavy refactoring on the unit tests. I ended up increasing the number of tests by nearly 33% (went from 29 tests before to 39 tests now), but also managed to reduce the number of lines by nearly half (was 942 lines before, now 449 lines). This was mostly just extracting duplicated setup code into the `beforeEach` methods, and leveraging some native `sinon` functionality on stubs
* I ended up going with an [implementation](https://github.com/swellaby/vsts-mirror-git-repository/pull/26/files#diff-1a7185c47edb4674d1cd78c50e05fd21R66) for excluding PR refs by simply nuking those refs after cloning. There's a few different approaches that could work, but this was the simplest/most straightforward option to plug in to the existing workflow, whereas the others would've required some additional changes.

File: src/git-mirror/git-mirror-task.ts [hunk 1]
```diff
@@ -115,7 +138,7 @@ export class GitMirrorTask {
         }    
     }
 
-    private taskIsRunnning(): number {
```
Comment (traviskosarek) [inline src/git-mirror/git-mirror-task.ts]: I can't believe I misspelled this :(, good catch

**Comment by calebcartwright** [comment]
So I actually started down the path described in the linked article, but I could not get it to work (even manually). In reading through some of the comments on that article, it seems others experienced the same problems.

The current implemented approach is nuking all refs that contain the substring `refs/pull` which covers both GitHubs pattern of `refs/pull...` and it will also cover Bitbuckets pattern of `refs/pull-request...` . The only way the current approach would _miss_ a PR ref is if some other SCM vendor does not start the ref name with `pull`. 

Like I said before, there's a lot of different ways to avoid attempting to push PR refs, but I still think this is the best approach. It's still doing a full mirror operation, which I think is important given that this extension is _named_ `git mirror` ;) , and it's already functional while excluding all known PR refs (have tested it with GitHub, Bitbucket, and Azure Repos already).

I'd vote that we go ahead and merge/publish with this solution given that it solves everything that we know today. If we discover something else down the road (idk perhaps GitLab uses different PR ref naming conventions) then we can cross that bridge when we get to it, but I don't think we should let this bug continue to live in the interim.

**Comment by traviskosarek** [comment]
I'm good with that. I just wanted to bring it up in case you see something I didn't in terms of future asks :)

To clarify on the linked article, I meant to say "follow the concept, not the implementation". I couldn't get their implementation to work manually either. What I was proposing as a possible alternative is just modifying what gets pushed vs deleting local (to the build) copies of refs that are known to be broken.

You can do a command like:

`git push source destination:ref/heads/*`

Where you specify the refspec on what gets pushed (in the above example "ref/heads/*")

I agree that the way you have it is probably good for 99%+ of all scenarios and that anything that comes up later is for future @calebcartwright and @traviskosarek to figure out, haha.

**Comment by calebcartwright** [comment]
I am all for punting any potential future changes to our future selves haha. I didn't realize this til i started running tests with some of our repos that have been around for a while, but this error would've popped up for anyone working with a repo that had ever had a pull request. Progress!

**Pull request merged**